### DISCUSSION TITLE
Location access on Android (ML2)

### DISCUSSION THREAD

**Discussion by paulmelis** [StereoKit/StereoKit#904] [Q&A]
I'm trying to get a (coarse) location on a Magic Leap 2, just once at application startup is enough. As far as I can tell one way to get location information is to use a [LocationManager](https://learn.microsoft.com/en-us/xamarin/android/platform/maps-and-location/location#location-manager) (or [here](https://developer.oculus.com/documentation/native/android/move-location/) for Quest 3) object. 

```
using Android.App;
using Android.Content;
using Android.Locations;
using Android.OS;
using Android.Runtime;

...
    // Called after SK.Initialize()
    GCHandle gch = GCHandle.FromIntPtr(SK.Settings.androidActivity);
    Context context = (Context)gch.Target;
    LocationManager location_manager = (LocationManager) context.GetSystemService(Context.LocationService);
...
```

This seems pretty straight-forward, but I get a crash (SIGABRT, looking at the logcat output) at the last line when launching the app on the ML2. 

I'm not very familiar with Android development, so I'm probably doing some stupid things. Yet, I'm wondering if the `androidActivity` is indeed what I think it is and if I'm using it in the right manner? Has someone done something similar on Android already?

**Comment by maluoi** [comment]
I'd probably not recommend using StereoKit's androidActivity, there's a couple different paths I use to collect the Activity, and it may not always be present. Not to mention the IntPtr thing 😅

I'd recommend using `Android.App.Application.Context` instead, this seems to be a pretty conveniently available global variable that's part of the official API, and it gets you straight to the `Context` here anyhow!

**Reply by paulmelis** [reply]
Actually, I tried that initially, but same result :cry:

**Comment by paulmelis** [comment]
After a couple of hours more trying to figure out what is wrong I'm not much further, unfortunately. Deploying to a Quest 3 gives the same SIGABRT, but slightly more info in the logcat output:

```
--------- beginning of crash
03-15 21:46:35.242  8053  8102 F libc    : Fatal signal 6 (SIGABRT), code -1 (SI_QUEUE) in tid 8102 (StereoKit Main), pid 8053 (d_location_test)
03-15 21:46:35.168     0     0 I CAM_INFO: CAM-SENSOR: cam_sensor_driver_cmd: 1265 20:46:35.147 CAM_START_DEV Success for imx471 sensor_id:0x471,sensor_slave_addr:0x34
03-15 21:46:35.335     0     0 I msm_cvp : sess: Opening cvp instance: 0000000000000000
03-15 21:46:35.337     0     0 I synx    : info: synx_initialize: 2287: [sess :9] session created cvp-kernel-client
03-15 21:46:35.345     0     0 I synx    : info: synx_initialize: 2287: [sess :10] session created umd-client-3262
03-15 21:46:35.854  8352  8352 F DEBUG   : *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
03-15 21:46:35.854  8352  8352 F DEBUG   : Build fingerprint: 'oculus/eureka/eureka:12/SQ3A.220605.009.A1/49872870047700510:user/release-keys'
03-15 21:46:35.854  8352  8352 F DEBUG   : Revision: '0'
03-15 21:46:35.854  8352  8352 F DEBUG   : ABI: 'arm64'
03-15 21:46:35.854  8352  8352 F DEBUG   : Timestamp: 2024-03-15 21:46:35.302935736+0100
03-15 21:46:35.854  8352  8352 F DEBUG   : Process uptime: 6s
03-15 21:46:35.854  8352  8352 F DEBUG   : Diagnostic-Collection-Type: Tombstone Proto
03-15 21:46:35.854  8352  8352 F DEBUG   : Has-Diagnostic-Summary: Yes
03-15 21:46:35.854  8352  8352 F DEBUG   : APK-Location: /data/app/~~u3w_lt4Plm3RgFRoafoBsw==/com.companyname.android_location_test-0Wqm1Ycyv8sY6szmeBbb4Q==/base.apk
03-15 21:46:35.854  8352  8352 F DEBUG   : Debugging: No
03-15 21:46:35.854  8352  8352 F DEBUG   : Has-Waited-For-Debugger: No
03-15 21:46:35.854  8352  8352 F DEBUG   : Last-Activity-Time: 1081479
03-15 21:46:35.854  8352  8352 F DEBUG   : Process-Runtime: 6258
03-15 21:46:35.854  8352  8352 F DEBUG   : Waiting-To-Kill: 
03-15 21:46:35.854  8352  8352 F DEBUG   : Killed-By-AM: No
03-15 21:46:35.854  8352  8352 F DEBUG   : Removed: No
03-15 21:46:35.854  8352  8352 F DEBUG   : Wanted-ANR-Ui: No
03-15 21:46:35.854  8352  8352 F DEBUG   : Killed: No
03-15 21:46:35.854  8352  8352 F DEBUG   : OOM-Score-Adj: 900
03-15 21:46:35.854  8352  8352 F DEBUG   : Wanted-Error-Ui: No
03-15 21:46:35.854  8352  8352 F DEBUG   : Kill-Time: 0
03-15 21:46:35.854  8352  8352 F DEBUG   : State: CRE 
03-15 21:46:35.854  8352  8352 F DEBUG   : Foreground: No
03-15 21:46:35.854  8352  8352 F DEBUG   : Executable: /system/bin/app_process64
03-15 21:46:35.854  8352  8352 F DEBUG   : Cmdline: com.companyname.android_location_test
03-15 21:46:35.854  8352  8352 F DEBUG   : pid: 8053, tid: 8102, name: StereoKit Main  >>> com.companyname.android_location_test <<<
03-15 21:46:35.854  8352  8352 F DEBUG   : uid: 10008
03-15 21:46:35.854  8352  8352 F DEBUG   : tagged_addr_ctrl: 0000000000000001
03-15 21:46:35.854  8352  8352 F DEBUG   : signal 6 (SIGABRT), code -1 (SI_QUEUE), fault addr --------
03-15 21:46:35.854  8352  8352 F DEBUG   : Abort message: 'No pending exception expected: java.lang.IllegalArgumentException: Receiver not registered: com.oculus.vrapi.SystemActivitiesReceiver@25c1256
03-15 21:46:35.854  8352  8352 F DEBUG   :   at android.content.IIntentReceiver android.app.LoadedApk.forgetReceiverDispatcher(android.content.Context, android.content.BroadcastReceiver) (LoadedApk.java:1582)
03-15 21:46:35.854  8352  8352 F DEBUG   :   at void android.app.ContextImpl.unregisterReceiver(android.content.BroadcastReceiver) (ContextImpl.java:1798)
03-15 21:46:35.854  8352  8352 F DEBUG   :   at void android.content.ContextWrapper.unregisterReceiver(android.content.BroadcastReceiver) (ContextWrapper.java:771)
03-15 21:46:35.854  8352  8352 F DEBUG   :   at void com.oculus.vrapi.SystemActivitiesReceiver.stopReceiver(android.content.Context) (:9)
```

Apparently somewhere an illegal argument is passed, causing an (unexpected) exception, causing the abort. This is with this simple standalone test code:

```
using StereoKit;
#if ANDROID
using Android.App;
using Android.Content;
using Android.Locations;
using Android.OS;
using Android.Runtime;
#endif

namespace android_location_test;

class Program : Fragment, ILocationListener
{
	static void Main(string[] args)
	{
		// Initialize StereoKit
		SKSettings settings = new SKSettings
		{
			appName = "android-location-test",
			assetsFolder = "Assets",
		};

		if (!SK.Initialize(settings))
			return;

        Program app = new Program();
        app.Run();

        SK.Shutdown();
    }

    public void OnLocationChanged(Location location)
    {
        Log.Info($"LOCATION {location}");
    }

    public void OnProviderDisabled(string provider) {}
    public void OnProviderEnabled(string provider) {}
    public void OnStatusChanged(string provider, [GeneratedEnum] Availability status, Bundle extras) {}

    public void Run()
    {
#if ANDROID
                Context context = Android.App.Application.Context;
                Log.Info($"context = {context}");
                LocationManager location_manager = (LocationManager) context.GetSystemService("location");
                Log.Info($"location manager = {location_manager}");
#endif

		// Create assets used by the app
		Pose  cubePose = new Pose(0, 0, -0.5f);
		Model cube     = Model.FromMesh(
			Mesh.GenerateRoundedCube(Vec3.One*0.1f, 0.02f),
			Material.UI);

		Matrix   floorTransform = Matrix.TS(0, -1.5f, 0, new Vec3(30, 0.1f, 30));
		Material floorMaterial  = new Material("floor.hlsl");
		floorMaterial.Transparency = Transparency.Blend;

		// Core application loop
                while (SK.Step(() =>
                {		
			        if (SK.System.displayType == Display.Opaque)
				        Mesh.Cube.Draw(floorMaterial, floorTransform);
        
			        UI.Handle("Cube", ref cubePose, cube.Bounds);
			        cube.Draw(cubePose.ToMatrix());
		        }));
	        }
}
```

Any tips appreciated

**Reply by maluoi** [reply]
I don't do a ton of code like this, so I may be missing something more obvious, but here are things I like to check when working with Android APIs:
- Do I have permissions? There's a collection of [location related permissions that may be required](https://learn.microsoft.com/en-us/xamarin/android/platform/maps-and-location/location#location-permissions).
- Is it enough to put those permissions in the manifest, or [do I need to prompt the user](https://github.com/StereoKit/StereoKit/blob/master/Examples/StereoKitTest/StereoKitTest_Xamarin/MainActivity.cs#L59-L64).
- Is the code thread sensitive? StereoKit does not run on the Activity's "main thread" in the template. Not a frequent thing, but I have seen this. `Activity.RunOnUiThread` can help in these cases.
- Double check manifest stuff again. It's almost always a manifest problem.

I'm not really seeing a lot of registration boilerplate code here, if this is all of your code related to it, then there's probably a lot missing you still need to do. The first link here may be of use if so.

**Reply by paulmelis** [reply]
Thanks for the pointers. It seems I'm hitting upon something more fundamental, probably not related to anything location-related. Even printing a simple `Android.App.Application.ProcessName` already gives a SIGABRT on Quest 3 (built and deployed from Linux, dotnet 7.0.115, SK 0.3.9, SK Templates 0.7.2, btw). I'll check if the same happens when building and deploying from Windows.

I mean, this should work, right? Or would there be a possible calling-thread issue when accessing the `ProcessName` property? Commenting out the Log.Info line makes it run without issue, leaving it in causes a SIGABRT.

```
using StereoKit;

namespace android_location_test;

class Program
{
	static void Main(string[] args)
	{
                // Initialize StereoKit
                SKSettings settings = new SKSettings
                {
	                appName = "android-location-test",
	                assetsFolder = "Assets",
                };
                
                if (!SK.Initialize(settings))
	                return;

                Program app = new Program();
                app.Run();
        
                SK.Shutdown();
    }

    public void Run()
    {
#if ANDROID
        Log.Info($"Android.App.Application.ProcessName = {Android.App.Application.ProcessName}");
#endif

        // Create assets used by the app
        Pose  cubePose = new Pose(0, 0, -0.5f);
        Model cube     = Model.FromMesh(
	        Mesh.GenerateRoundedCube(Vec3.One*0.1f, 0.02f),
	        Material.UI);
        
        Matrix   floorTransform = Matrix.TS(0, -1.5f, 0, new Vec3(30, 0.1f, 30));
        Material floorMaterial  = new Material("floor.hlsl");
        floorMaterial.Transparency = Transparency.Blend;
        
        // Core application loop
        while (SK.Step(() =>
        {		
	        if (SK.System.displayType == Display.Opaque)
		        Mesh.Cube.Draw(floorMaterial, floorTransform);
        
	        UI.Handle("Cube", ref cubePose, cube.Bounds);
	        cube.Draw(cubePose.ToMatrix());
        }));
    }
}
```

**Reply by maluoi** [reply]
Ah interesting! I just deployed this code from Windows to my Quest3, and it worked fine for me. I am using .NET SDK 8.0.103, but otherwise the same.

**Reply by paulmelis** [reply]
So first piece of the puzzle is the cause of the SIGABRT I'm seeing. Apparently there's an obscure bug when an APK gets incrementally built, causing inclusion of the wrong DLL (https://github.com/xamarin/xamarin-android/issues/7580#issue-1461831953). Work-around is to remove the `Projects/Android/obj` dir before building. With this fix at least the SIGBART goes away for me, even when using dotnet on Linux

Edit: the linked bug might not be entirely the correct one, I see similar stuff in the xamarin-android issue list, but the work-around there does work

**Reply by paulmelis** [reply]
Regarding permissions and the manifest, I've now added this to `AndroidManifest.xml`:

```
...
        <service android:name="MyNagivationService" android:foregroundServiceType="location">
        </service>
    </application>

    <uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION" />
```

The `<service>` part is apparently needed, according to https://developer.android.com/develop/sensors-and-location/location/permissions#foreground

As far as the runtime check/request for permissions, I'm using this for now:

```
        Context context = Android.App.Application.Context;
        if (context.CheckSelfPermission(Manifest.Permission.AccessCoarseLocation) == Android.Content.PM.Permission.Granted)
        {
            LocationManager location_manager = (LocationManager) context.GetSystemService(Context.LocationService);
            foreach (string p in location_manager.GetProviders(false))
                Log.Info($"LOCATION PROVIDER {p}");

            Location location = location_manager.GetLastKnownLocation("fused");
            if (location != null)
                Log.Info($"LAST KNOWN LOCATION (fused) = lat {location.Latitude}, lon {location.Longitude}, alt {location.Altitude}");
        }
        else
        {
            Log.Warn("NO PERMISSION FOR COARSE LOCATION ACCESS");
            // ...RequestPermissions(new string[] { Manifest.Permission.AccessCoarseLocation }, 1);
        }
```

The problem is I don't know on which instance to call `RequestPermissions()`. The SK code [you linked](https://github.com/StereoKit/StereoKit/blob/master/Examples/StereoKitTest/StereoKitTest_Xamarin/MainActivity.cs#L59-L64) seems to use the `AppActivityCompat` class from package `AndroidX.AppCompat.App` to bring in `RequestPermissions()`. But as far as I can tell on NuGet that package is [only available](https://www.nuget.org/packages/Xamarin.AndroidX.AppCompat/) for .NET 6.0?

**Reply by paulmelis** [reply]
Btw, it seems I do need `Activity.RunOnUiThread` for some of the next calls I need. But I don't see how I can easily get to the `Activity` instance? Is of the SK IntPtrs usable at all, or should I go a different route?

**Reply by maluoi** [reply]
It's possible that SK's `IntPtr` might be enough, but I'm not good enough at interop with Java objects to be really confident about getting it right! I just do a really simple global `static`/singleton thing if I need the actual Activity:

```csharp
public class MainActivity : Activity, ISurfaceHolderCallback2
{
	View surface;

	public static Activity Activity { get; private set; }

	protected override void OnCreate(Bundle savedInstanceState)
	{
		Activity = this;
		base.OnCreate(savedInstanceState);
		Run();
		SetContentView(Resource.Layout.activity_main);
	}
	// ...
```
